pkg/exporter/syslogexporter/syslog.go

```diff
@@ -0,0 +1,206 @@
+		logger:                   logger,
+		hostname:                 hostname,
+		network:                  cfg.Protocol,
+		addr:                     fmt.Sprintf("%s:%d", cfg.Endpoint, cfg.Port),
+		format:                   cfg.Format,
+		tlsConfig:                tlsConfig,
+		pid:                      pid,
+		app:                      app,
+		additionalStructuredData: cfg.AdditionalStructuredData,
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.connect()
+	if err != nil {
+		return nil, err
+	}
+	return s, err
+}
+
+func (s *Syslog) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.conn != nil {
+		err := s.conn.Close()
+		s.conn = nil
+		return err
+	}
+	return nil
+}
+
+func (s *Syslog) connect() error {
+	if s.conn != nil {
+		s.conn.Close()
+		s.conn = nil
+	}
+	var err error
+	if s.tlsConfig != nil {
+		s.conn, err = tls.Dial("tcp", s.addr, s.tlsConfig)
+	} else {
+		s.conn, err = net.Dial(s.network, s.addr)
+	}
+	return err
+}
+
+func (s *Syslog) Write(msg map[string]any) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	/* isFormatCorrect := s.validateFormat(msg)
+
+	if !isFormatCorrect && s.dropInvalidMsg {
+		s.logger.Debug("Invalid message format",
+			zap.String("format", s.format),
+			zap.String("msg", msg),
+		)
+		return nil
+	} else if !isFormatCorrect {
+		msg = s.formatMsg(msg)
+	}*/
+	s.addStructuredData(msg)
+	msgStr := s.formatMsg(msg)
+
+	if s.conn != nil {
+		if err := s.write(msgStr); err == nil {
+			return nil
+		}
+	}
+	if err := s.connect(); err != nil {
+		return err
+	}
+
+	return s.write(msgStr)
+}
+
+func (s *Syslog) write(msg string) error {
+	// check if logs contains new line character at the end, if not add it
+	if !strings.HasSuffix(msg, "\n") {
+		msg = fmt.Sprintf("%s%s", msg, "\n")
+	}
+	_, err := fmt.Fprint(s.conn, msg)
+	return err
+}
+
+func (s *Syslog) formatMsg(msg map[string]any) string {
+	switch s.format {
+	case formatRFC3164Str:
+		return formatRFC3164(msg)
+	case formatRFC5424Str:
+		return formatRFC5424(msg)
+	default:
+		return ""
+	}
+}
+
+func (s *Syslog) addStructuredData(msg map[string]any) {
+	if len(s.additionalStructuredData) == 0 {
+		return
+	}
+	_, ok := msg[structuredData]
+	if !ok {
+		msg[structuredData] = s.additionalStructuredData
+	}
+}
+
+func populateDefaults(msg map[string]any, msgProperty string) {
+	const emptyValue = "-"
+	msgValue, ok := msg[msgProperty]
+	if !ok && msgProperty == priority {
+		msg[msgProperty] = defaultPriority
+		return
+	}
+	if !ok && msgProperty == version {
+		msg[msgProperty] = versionRFC5424
+		return
+	}
+	if !ok && msgProperty == facility {
+		msg[msgProperty] = defaultFacility
+		return
+	}
+	if !ok {
+		msg[msgProperty] = emptyValue
+		return
+	}
+	msg[msgProperty] = msgValue
+}
+
+func formatRFC3164(msg map[string]any) string {
+	// timestamp := time.Now().Format(time.Stamp)
+	msgProperties := []string{priority, hostname, message}
+	for _, msgProperty := range msgProperties {
+		populateDefaults(msg, msgProperty)
+	}
```

Comment:
I would move iteration to the `populateDefaults` function

Reply:
done